### PR TITLE
feat(gatsby): enable concurrent features

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

React is moving forward with concurrent mode. It's going to rebrand it to React concurrent features. We'll be adding early support for it.

Swapping `ReactDOM.render` to `ReactDOM.unstable_createRoot`. It should be 99% equal to regular hydrate but some behavior changes might occur.

There is no hydration callback anymore so I swapped it to `onMount`.

### CODE REVIEW

File: packages/gatsby/src/utils/flags.ts [hunk 1]
```diff
@@ -176,6 +176,16 @@ const activeFlags: Array<IFlag> = [
...
     testFitness: (): fitnessEnum => true,
   },
+  {
+    name: `CONCURRENT_FEATURES`,
+    env: `GATSBY_CONCURRENT_FEATURES`,
+    command: `all`,
+    telemetryId: `ConcurrentFeatures`,
+    experimental: true,
+    description: `Enable React's concurrent features`,
+    // umbrellaIssue: `https://gatsby.dev/concurrent-features`,
```
Comment (wardpeet) [inline packages/gatsby/src/utils/flags.ts:186]: Disabled it for now as we don't want to expose it to much

File: packages/gatsby/cache-dir/production-app.js [hunk 2]
```diff
@@ -175,24 +175,42 @@ apiRunnerAsync(`onClientEntry`).then(() => {
...
-      )
+          : null
+
+      if (renderer === ReactDOM.unstable_createRoot) {
+        renderer(container, {
+          hydrate: true,
+        }).render(<App />)
+      } else {
+        renderer(<App />, container)
+      }
```
Comment (wardpeet) [inline packages/gatsby/cache-dir/production-app.js:213]: concurrent mode does not have a callback anymore so switched to onMount inside app

File: packages/gatsby/src/utils/flags.ts [hunk 3]
```diff
@@ -176,6 +176,30 @@ const activeFlags: Array<IFlag> = [
     umbrellaIssue: `https://gatsby.dev/functions-feedback`,
     testFitness: (): fitnessEnum => true,
   },
+  {
+    name: `CONCURRENT_FEATURES`,
+    env: `GATSBY_CONCURRENT_FEATURES`,
```
Comment (vladar) [inline packages/gatsby/src/utils/flags.ts:181]: Should it be `GATSBY_EXPERIMENTAL_CONCURRENT_FEATURES` for consistency?

File: packages/gatsby/cache-dir/production-app.js [hunk 4]
```diff
@@ -175,24 +175,42 @@ apiRunnerAsync(`onClientEntry`).then(() => {
...
+      }, [])
+
+      return <GatsbyRoot>{SiteRoot}</GatsbyRoot>
+    }
 
     const renderer = apiRunner(
       `replaceHydrateFunction`,
       undefined,
-      ReactDOM.hydrate
+      process.env.GATSBY_CONCURRENT_FEATURES
```
Comment (LekoArts) [inline packages/gatsby/cache-dir/production-app.js:196]:
```suggestion
      process.env.GATSBY_EXPERIMENTAL_CONCURRENT_FEATURES
```

**Review by LekoArts — APPROVED** [review]
👍

**Pull request merged**